---
title: "phylogeny_renaming"
author: "Brenda_Kiage"
date: "2023-07-17"
output: html_document
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

```


```{r read in the required libraries}

library(readr)
library(dplyr)
library(Biostrings)
library(phylotools)
library(tidyr)
library(readxl)

```



```{r Read in the results/metadata excel sheet}
flub_df <- read_xlsx("~/Downloads/flub_facility.xlsx")
flub_df$sample_id <- gsub("A50115", "A50015", flub_df$sample_id) 

```


```{r Read in the concatenated sequences}

NA. <- readDNAStringSet("~/Desktop/Msc/flub_results/concatenated/NA_concat.fasta")
NA_df <- as.data.frame(NA.)
HA <- readDNAStringSet("~/Desktop/Msc/flub_results/concatenated/HA_concat.fasta")
HA_df <- as.data.frame(HA)
PB1 <- readDNAStringSet("~/Desktop/Msc/flub_results/concatenated/PB1_concat.fasta")
PB1_df <- as.data.frame(PB1)
PB2 <- readDNAStringSet("~/Desktop/Msc/flub_results/concatenated/PB2_concat.fasta")
PB2_df <- as.data.frame(PB2)
PA <- readDNAStringSet("~/Desktop/Msc/flub_results/concatenated/PA_concat.fasta")
PA_df <- as.data.frame(PA)
NP <- readDNAStringSet("~/Desktop/Msc/flub_results/concatenated/NP_concat.fasta")
NP_df <- as.data.frame(NP)
MP <- readDNAStringSet("~/Desktop/Msc/flub_results/concatenated/MP_concat.fasta")
MP_df <- as.data.frame(MP)
NS <- readDNAStringSet("~/Desktop/Msc/flub_results/concatenated/NS_concat.fasta")
NS_df <- as.data.frame(NS)


#Extract the names of sequences and generate a dataframe

NA_names <- names(NA.)
NA_names_df <- as.data.frame(NA_names)
HA_names <- names(HA)
HA_names_df <- as.data.frame(HA_names)
PB1_names <- names(PB1)
PB1_names_df <- as.data.frame(PB1_names)
PB2_names <- names(PB2)
PB2_names_df <- as.data.frame(PB2_names)
PA_names <- names(PA)
PA_names_df <- as.data.frame(PA_names)
NP_names <- names(NP)
NP_names_df <- as.data.frame(NP_names)
MP_names <- names(MP)
MP_names_df <- as.data.frame(MP_names)
NS_names <- names(NS)
NS_names_df <- as.data.frame(NS_names)


```


```{r Bind the two dataframes}
NS_df1 <- cbind(NS_names_df, NS_df)
colnames(NS_df1) <- c("sample_id", "sequences")

#Left join the two dataframes

NS_df1 <- merge(NS_df1,flub_df %>% select('sample_id','facility_name',"Collection_date"), by="sample_id", all.x=TRUE)

```



```{r  Select the complete genomes for further analysis}

completeNS <- read.delim("~/Desktop/Msc/flub_results/completeNS.txt", header = FALSE)
colnames(completeNS) <- "sample_id"

complete_df <-  read.delim("~/Desktop/Msc/flub_results/complete.txt", header = FALSE)
colnames(complete_df) <- "sample_id"
complete_samples <- complete_df[-c(1,16,32,35,47),]
complete_samples <- as.data.frame(complete_samples)
colnames(complete_samples) <- "sample_id"

NS_df1 <- subset(NS_df1, sample_id %in% completeNS$sample_id)
NS_df2 <- subset(NS_df1, sample_id %in% complete_samples$sample_id)


NS_df2 <- unite(NS_df2, phylogeny_name, c(sample_id,facility_name,Collection_date), sep = '|')

write_csv(NS_df2, "~/Desktop/flub_msc/results/NS_data.csv")
```


```{r rename the sequences with NCBI ids}
NCBI_names <- read_csv("~/Downloads/HA_NCBI_names.csv")
NCBI_names$name <- gsub(">", "", NCBI_names$name)
colnames(NCBI_names) <- c("phylogeny_name","NCBI_id")
NS_df2 <- merge(NS_df2,NCBI_names %>% select('phylogeny_name','NCBI_id'), by="phylogeny_name", all.x=TRUE)


```


```{r Write outputs to a fasta file }

writeFasta<-function(data, filename){
  fastaLines = c()
  for (rowNum in 1:nrow(data)){
    fastaLines = c(fastaLines, as.character(paste(">", data[rowNum,"NCBI_id"], sep = "")))
    fastaLines = c(fastaLines,as.character(data[rowNum,"sequences"]))
  }
  fileConn<-file(filename)
  writeLines(fastaLines, fileConn)
  close(fileConn)
}

writeFasta(NS_df2, '~/Desktop/Msc/flub_results/renamed/NS_renamed_complete.fa')

```


```{r renaming of reference sequences for concatenated trees}

NA. <- readDNAStringSet("~/Desktop/flub_msc/results/mafft/NA_edit.fa")
NA_df <- as.data.frame(NA.)
HA <- readDNAStringSet("~/Desktop/flub_msc/results/mafft/HA_edit.fa")
HA_df <- as.data.frame(HA)
PB1 <- readDNAStringSet("~/Desktop/flub_msc/results/mafft/PB1_edit.fa")
PB1_df <- as.data.frame(PB1)
PB2 <- readDNAStringSet("~/Desktop/flub_msc/results/mafft/PB2_edit.fa")
PB2_df <- as.data.frame(PB2)
PA <- readDNAStringSet("~/Desktop/flub_msc/results/mafft/PA_edit.fa")
PA_df <- as.data.frame(PA)
NP <- readDNAStringSet("~/Desktop/flub_msc/results/mafft/NP_edit.fa")
NP_df <- as.data.frame(NP)
MP <- readDNAStringSet("~/Desktop/flub_msc/results/mafft/MP_edit.fa")
MP_df <- as.data.frame(MP)
NS <- readDNAStringSet("~/Desktop/flub_msc/results/mafft/NS_edit.fa")
NS_df <- as.data.frame(NS)


#Extract the names of sequences and generate a dataframe

NA_names <- names(NA.)
NA_names_df <- as.data.frame(NA_names)
HA_names <- names(HA)
HA_names_df <- as.data.frame(HA_names)
PB1_names <- names(PB1)
PB1_names_df <- as.data.frame(PB1_names)
PB2_names <- names(PB2)
PB2_names_df <- as.data.frame(PB2_names)
PA_names <- names(PA)
PA_names_df <- as.data.frame(PA_names)
NP_names <- names(NP)
NP_names_df <- as.data.frame(NP_names)
MP_names <- names(MP)
MP_names_df <- as.data.frame(MP_names)
NS_names <- names(NS)
NS_names_df <- as.data.frame(NS_names)



```


```{r}
NS_df1 <- cbind(NS_names_df, NS_df)
colnames(NS_df1) <- c("sample_id", "sequences")


vic_references <- NS_df1[grepl("Victoria", NS_df1$sample_id), ]
ref_df <- as.data.frame(vic_references)
ref_df <- ref_df %>% separate(sample_id, c("sample_id","Lineage", "segment", "clade", "collection_date"), sep = "\\|")
ref_df <- ref_df %>% separate(collection_date, c("year", "month", "date"), sep = "-")
ref_df <- ref_df %>% unite(sample_id, c(sample_id, Lineage, clade), sep = "|")

yam_references <- NS_df1[grepl("Yamagata", NS_df1$sample_id), ]
yam_ref_df <- as.data.frame(yam_references)
yam_ref_df <- yam_ref_df %>% separate(sample_id, c("sample_id","Lineage", "segment", "clade", "collection_date"), sep = "\\|")
yam_ref_df <- yam_ref_df %>% separate(collection_date, c("year", "month", "date"), sep = "-")
yam_ref_df <- yam_ref_df %>% unite(sample_id, c(sample_id, Lineage, clade), sep = "|")

samples <- NS_df1[!grepl("Yamagata", NS_df1$sample_id), ]
samples_df <- as.data.frame(samples)
samples1 <- samples_df[!grepl("Victoria", samples_df$sample_id), ]
samples_df <- as.data.frame(samples1)

concat_df <- bind_rows(ref_df, yam_ref_df, samples_df)

writeFasta<-function(data, filename){
  fastaLines = c()
  for (rowNum in 1:nrow(data)){
    fastaLines = c(fastaLines, as.character(paste(">", data[rowNum,"sample_id"], sep = "")))
    fastaLines = c(fastaLines,as.character(data[rowNum,"sequences"]))
  }
  fileConn<-file(filename)
  writeLines(fastaLines, fileConn)
  close(fileConn)
}

writeFasta(concat_df, '~/Desktop/flub_msc/results/mafft/NS_concat.fa')

```

